posts/2020-02-12-using-dotty-union-types-with-akka-typed-part-II.adoc

```diff
@@ -0,0 +1,285 @@
+----
+abstract class Behavior[T](private[akka] val _tag: Int) { behavior =>
+
+  /**
+   * Narrow the type of this Behavior, which is always a safe operation. This
+   * method is necessary to implement the contravariant nature of Behavior
+   * (which cannot be expressed directly due to type inference problems).
+   */
+  final def narrow[U <: T]: Behavior[U] = this.asInstanceOf[Behavior[U]]
+  
+  ...
+}
+  
+abstract class ExtensibleBehavior[T] extends Behavior[T](BehaviorTags.ExtensibleBehavior) {
+
+  ...
+  
+  def receive(ctx: TypedActorContext[T], msg: T): Behavior[T]
+  
+  ...
+----
+
+There's quite a bit going on here.
+
+First, we note that the `Behavior` class is generic: it has a type parameter `T`, which, because of no specific variance annotation on `T`, implies that `Behavior` is _invariant_ in its type parameter `T`. Also note the comment on the `narrow` method, stating: 
+
+> _This method is necessary to implement the contravariant nature of Behavior (which cannot be expressed directly due to type inference problems)_.
+
+Second, we see that the class `ExtensibleBehavior`, which is a subclass of `Behavior`, has a `receive` method which takes a parameter `msg` of type `T`. Because functions (or _methods_), are contravariant in their argument types, the only possible variance case for the type parameter is invariant (`T`) or contravariant (`-T`). Because of type inference problems in Scala 2, the former was chosen.
+
+NOTE: Variance manifests itself in specific contexts and is a topic in itself with _contravariance_ being the least intuitive. We'll see however that, in the case of `Behavior`, it is actually quite easy to understand. I'll get back to that later. For a comprehensive explanation of variance in Scala read https://docs.scala-lang.org/tour/variances.html[this article].
+
+Finally, we see from the definition of the `narrow` method, that it returns a behavior which is more restrictive in its type than the behavior on which it is called. The implementation of `narrow` uses `asInstanceOf` to apply this restriction. 
+
+== Clearing up some magic in the context of Dotty
+
+All the above is nice, but it will leave some readers with questions. So, let's look at this from a practical point of view by looking at the Dotty version of the `Pinger` which uses Union types.
+
+Starting from the (internal) protocol definition:
+
+[source,scala,linenums]
+----
+object Pinger {
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+
+  // My protocol + the responses I need to understand...
+  type CommandsAndResponses = Command | PingPong.Response
+}
+
+object PingPong {
+  sealed trait Command
+  final case class Ping(replyTo: ActorRef[Response]) extends Command
+
+  sealed trait Response
+  case object Pong extends Response
+}
+----
+
+We can run the following experiment (in _dotr_, the Dotty REPL):
+
+[source,scala,linenums]
+----
+scala> import akka.actor.typed.ActorRef
```

Comment:
haha this block would be a **perfect** use case for mdoc 😉 . After reading a couple of these, I'm really curious how hard it would be for it to support `adoc` files. It may not be at all. I'll explore and bug Olaf about it.

Reply:
For anyone reading this and wondering what mdoc is --> https://github.com/scalameta/mdoc